Rules/UseProcessBlockForPipelineCommands.cs

```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+#if !CORECLR
+    [Export(typeof(IScriptRule))]
+#endif
+    public class UseProcessBlockForPipelineCommands : IScriptRule
+    {
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            IEnumerable<Ast> functionAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+
+            foreach (FunctionDefinitionAst funcAst in functionAsts)
+            {
+                if
+                (
+                    funcAst.Body != null
+                    && funcAst.Body.ParamBlock != null
+                    && funcAst.Body.ParamBlock.Attributes != null
+                    && funcAst.Body.ParamBlock.Parameters != null
+                    && funcAst.Body.ProcessBlock == null
+                )

```

Comment:
You have quite a few indentation layers. You can eliminate some of them in some foreach loops by applying the following refactoring that makes the foreach loop continue if a condition is not met to avoid nesting the critical path in an if condition that leads to more indentation
```csharp
foreach(...)
{
    if (condition)
    {
        ...
    }
}
```
becomes
```csharp
foreach(...)
{
    if (!condition)
    {
        continue;
    }
    ...
}
```

Reply:
This should be resolved in dd428297d55f7256d4350133d64161b2679ced68. I took your advice and execution should drop out of each loop if conditions do not match, which keeps the return statement out of the if blocks.

Also, I realized that two of the conditional checks were not necessary:
### Check for null body

```csharp
// before
if
(
        funcAst.Body == null
	|| funcAst.Body.ParamBlock == null
	|| funcAst.Body.ParamBlock.Attributes == null
	|| funcAst.Body.ParamBlock.Parameters == null
	|| funcAst.Body.ProcessBlock != null
)

// after
if
(
	funcAst.Body.ParamBlock == null
	|| funcAst.Body.ParamBlock.Attributes == null
	|| funcAst.Body.ParamBlock.Parameters == null
	|| funcAst.Body.ProcessBlock != null
)
```

 I think checking for a null function body was wrong - I'd still want to warn on a pipeline parameter even if the body hadn't been implemented yet.

### Type check for AttributeAst
```csharp
// removed
if (paramAstAttribute is AttributeAst) { continue; }
```

This check was extraneous.